src/addon-loader.js

```diff
@@ -105,32 +80,47 @@ async function loadAddon(addonPath, verbose) {
     newSettings.moziot.config = {};
   }
 
-  return pluginClientPromise.then(() => {
-    console.log('Loading add-on for', manifest.name, 'from', addonPath);
+  const pluginClient = new PluginClient(manifest.name, {verbose});
 
-    let addonLoader;
-    try {
-      addonLoader = dynamicRequire(addonPath);
-      addonLoader(addonManagerProxy, newSettings, (packageName, errorStr) => {
-        console.error('Failed to load', packageName, '-', errorStr);
-        const message =
-          `Failed to start ${manifest.display_name} add-on: ${errorStr}`;
-        fail(message);
-      });
 
-      if (config.get('ipc.protocol') !== 'inproc') {
-        pluginClient.on('unloaded', () => {
-          sleep(500).then(() => process.exit(0));
-        });
-      }
-    } catch (e) {
+  return pluginClient.register()
+    .catch((e) => {
       console.error(e);
-      const message =
+      const err = `Failed to register package: ${manifest.name} with gateway`;
+      return Promise.reject(err);
+    })
+    .then((addonManagerProxy) => {
+      console.log('Loading add-on for', manifest.name, 'from', addonPath);
+      try {
+        const addonLoader = dynamicRequire(addonPath);
+        addonLoader(addonManagerProxy, newSettings, (packageName, errorStr) => {
+          console.error('Failed to load', packageName, '-', errorStr);
+          const message =
+          `Failed to start ${manifest.display_name} add-on: ${errorStr}`;
+          fail(addonManagerProxy, message);
+        });
+
+        if (config.get('ipc.protocol') !== 'inproc') {
+          pluginClient.on('unloaded', () => {
+            sleep(500).then(() => process.exit(0));
+          });
+        }
+      } catch (e) {
+        console.error(e);
+        const message =
         `Failed to start ${manifest.display_name} add-on: ${
           e.toString().replace(/^Error:\s+/, '')}`;
-      return fail(message);
-    }
-  });
+        return fail(addonManagerProxy, message);
```

Comment:
nit: fail doesn't return (it exits the process), so the return statement is never used.

Reply:
I've wondered about that too, but I didn't want to change the existing logic.

Reply:
@dhylands removed it